### PR TITLE
Support resource generation for SingleFile with code generator plugins

### LINKED ISSUE #34152: Support adding resources for single files with code generator compiler plugins
**Description:**
$subject

### LINKED ISSUE #34159: Bug in diagnostic reporting when compiling a module with a syntax error for the second time
**Description:**

When a new document is added to a module that has syntax errorsform an existing file, the expected diagnostic in the existing document is not reported.

**Steps to reproduce:**

Try the below test case
1 Create a project and add following content to main.bal:
```bal
public function main() {
}
; // expected syntax error - invalid token ";"
```
2. Write the below test case and execute:
```java
 @Test
  public void temp() {
     // load the package, compile and verify diagnostics
     Project project = ProjectLoader.loadProject(Paths.get(<bal-project-with-syntax-error>));
     PackageCompilation compilation1 = project.currentPackage().getCompilation();
     Assert.assertEquals(compilation1.diagnosticResult().errorCount(), 1);

     // Add another bal document (this doesn't have any errors)
     DocumentId documentId = DocumentId.create("foo.bal", project.currentPackage().getDefaultModule().moduleId());
     DocumentConfig documentConfig = DocumentConfig.from(documentId, "public function hello() {}", "foo.bal");
     Module newModule = project.currentPackage().getDefaultModule().modify().addDocument(documentConfig).apply();
        
     // Recompile and verify diagnostics
     PackageCompilation compilation2 = project.currentPackage().getCompilation();
     Assert.assertEquals(compilation2.diagnosticResult().errorCount(), 1); // failing
    }
```
**Affected Versions:**
SL Beta.5 at least

### PR DESCRIPTION
**Pull request by azinneera**
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGeneratorManager.java [hunk 1]
```diff
@@ -277,7 +277,9 @@ public void addSourceFile(TextDocument textDocument, String filenamePrefix, Modu
 
         @Override
         public void addSourceFile(TextDocument textDocument, String filenamePrefix) {
-            addSourceFile(textDocument, filenamePrefix, defaultModuleId);
+            if (this.currentPackage.project().kind().equals(ProjectKind.BUILD_PROJECT)) {
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGeneratorManager.java:280]: Do we need to inform the user here that source files cannot be added to single file project?

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGeneratorManager.java:280]: We can report a WARNING if it is a single file

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGeneratorManager.java]: Addressed in 2db6400 to show a warning as follows:
```
Compiling source
	hello_world.bal
WARNING [:(1:1,1:1)] skipping addition of generated file dummyfunc. sourcefile generation is not supported with standalone bal files

Generating executable
	bar.jar
```

**Review by pubudu91 — COMMENTED** [review]
The change related to the `DocumentContext` LGTM.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGeneratorManager.java [hunk 2]
```diff
@@ -267,6 +275,15 @@ public PackageCompilation compilation() {
 
         @Override
         public void addSourceFile(TextDocument textDocument, String filenamePrefix, ModuleId moduleId) {
+            if (this.currentPackage.project().kind().equals(ProjectKind.SINGLE_FILE_PROJECT)) {
+                DiagnosticInfo diagnosticInfo = new DiagnosticInfo(
+                        ProjectDiagnosticErrorCode.UNSUPPORTED_COMPILER_PLUGIN_TYPE.diagnosticId(),
+                        "skipping addition of generated file '" + filenamePrefix +
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGeneratorManager.java:281]: How about the following error message? 

"Skipped adding the generated source file with prefix "dummyfunc". Source file generation is not supported with standalone bal files"

**Review by sameerajayasoma — APPROVED** [review]

**Pull request merged**